sor/src/main/java/com/bazaarvoice/emodb/sor/audit/s3/AthenaAuditWriter.java

```diff
@@ -0,0 +1,502 @@
+        _logFilePrefix = requireNonNull(logFilePrefix, "logFilePrefix");
+        _clock = requireNonNull(clock, "clock");
+
+        // Audit queue isn't completely unbounded but is large enough to ensure at several times the normal write rate
+        // it can accept audits without blocking.
+        _auditQueue = new ArrayBlockingQueue<>(4096);
+
+        // Need to ensure the object mapper keeps the file stream open after each audit is written.
+        _objectWriter = objectMapper.copy()
+                .configure(JsonGenerator.Feature.AUTO_CLOSE_TARGET, false)
+                .writer();
+
+        // Two threads for the audit service: once to drain queued audits and one to close audit logs files and submit
+        // them for transfer.  Normally these are initially null and locally managed, but unit tests may provide
+        // pre-configured instances.
+        _auditService = Optional.ofNullable(auditService)
+                .orElse(Executors.newScheduledThreadPool(2, new ThreadFactoryBuilder().setNameFormat("audit-log-%d").build()));
+        _fileTransferService = Optional.ofNullable(fileTransferService)
+                .orElse(Executors.newSingleThreadExecutor(new ThreadFactoryBuilder().setNameFormat("audit-transfer-%d").build()));
+
+
+        long now = _clock.millis();
+        long msToNextBatch = _maxBatchTimeMs - (now % _maxBatchTimeMs);
+
+        // Do a one-time closing of all orphaned log files which may have been left behind if a previous EmoDB process
+        // terminated before closing them.
+        for (final File logFile : _stagingDir.listFiles((dir, name) -> name.startsWith(_logFilePrefix) && name.endsWith(OPEN_FILE_SUFFIX))) {
+            if (logFile.length() > 0) {
+                try {
+                    renameClosedLogFile(logFile);
+                } catch (IOException e) {
+                    _log.warn("Failed to close orphaned audit log file: {}", logFile, e);
+                }
+            } else {
+                if (!logFile.delete()) {
+                    _log.debug("Failed to delete empty orphaned log file: {}", logFile);
+                }
+            }
+        }
+
+        _auditService.scheduleWithFixedDelay(this::processQueuedAudits,
+                0, 1, TimeUnit.SECONDS);
+
+        _auditService.scheduleAtFixedRate(this::doLogFileMaintenance,
+                msToNextBatch, _maxBatchTimeMs, TimeUnit.MILLISECONDS);
+
+        requireNonNull(gracefulShutdownRegistry).registerWriteDependendent(this);
+    }
+
+    @Override
+    public void close() throws IOException {
+        _auditService.shutdown();
+        try {
+            if (!_auditService.awaitTermination(15, TimeUnit.SECONDS)) {
+                _log.warn("Audits still processing unexpectedly after shutdown");
+            } else {
+                // Poll the queue one last time and drain anything that is still remaining
+                processQueuedAudits();
```

Comment:
you are correct in that it will not work. I see no reason why we can't remove the `isShutdown()` check though. If it is still draining then if should just be allowed to continue.